### PR TITLE
Storage miner integration

### LINKED ISSUE #629: Find alternative WebSocket Library

### LINKED ISSUE #630: Implement Websocket Support

### LINKED ISSUE #631: Implement Websocket API calls

### LINKED ISSUE #632: Use Local Devnet to Test Storage Miner

### PR DESCRIPTION
**Pull request by AshantiMutinta**
**Summary of changes**
Changes introduced in this pull request:
- Adds websockets support and implements missing functions

**Reference issue to close (if applicable)**

Closes #629 
Closes #630 
Closes #631 
Closes #632 

**Other information and links**

### CODE REVIEW

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:217]: A task will be spawned for every function call of this, that won't ever exit. This should probably be one task where the receiver is cloned and passed back in this function?

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:684]: This will not match the Lotus implementation, are you sure you verified against? Enum representations are not internally tagged by default, and the keys are lower case, not first character capitalized.

Also for this, cloning a tipset just to serialize JSON is unnecessary, should take a reference to the tipset in the structure (if that means separating into two if both serialize and deserialize are needed, then it's worth it)

File: blockchain/chain_sync/src/sync.rs
Comment (austinabell) [inline blockchain/chain_sync/src/sync.rs:71]: is it possible to just wrap what is needed in an Arc<RwLock<..>> instead of wrapping the whole chain store in a mutex? This kinda worries me

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:271]: can you have the function you copied this from call this so we don't have duplication here?

File: forest/src/daemon.rs
Comment (austinabell) [inline forest/src/daemon.rs:95]: Was this a debugging log that wasn't removed? Seems duplicate and unnecessary

File: blockchain/chain/src/store/chain_store.rs
Comment (ec2) [reply blockchain/chain/src/store/chain_store.rs:217]: I think ashanti is mostly correct here, except for the fact that the loop never stops. You need to stop the loop once the receiver is dropped, i.e. the websocket connection is terminated. I think here, you could use the futures channels to better handle disconnects https://docs.rs/futures/0.3.5/futures/channel/mpsc/index.html. In async-std channels, i think you only detect a disconnection of all the receivers if the channel is full AND all recievers have been dropped.

File: node/rpc/src/lib.rs
Comment (ec2) [inline node/rpc/src/lib.rs:200]: you mean response?

File: node/rpc/src/lib.rs
Comment (ec2) [inline node/rpc/src/lib.rs:210]: is this necessarily a fail? this happens at disconnection too right?

**Comment by ec2** [comment]
Gunna go much more in depth tomorrow. My cloud lotus node is currently syncing.

File: blockchain/chain_sync/src/sync.rs
Comment (AshantiMutinta) [reply blockchain/chain_sync/src/sync.rs:71]: I thought about that but I think it makes sense in the context of both places being able to reflect the changes that are coming in from both sides. Like unless we go through the route of generating a new chain_store per operation which might take a lot of restructuring. Example things like HeadChange is something that would need to be reflected from both sides.

Additionally, I thought it was okay because state is behind one and trying to circumvent by using module level functions, getting the block_store from the state_manager and using sync types of the async library. 

I will look closer and see if we can just get what we need through

File: blockchain/chain/src/store/chain_store.rs
Comment (AshantiMutinta) [reply blockchain/chain/src/store/chain_store.rs:217]: nice! Shouldn't be hard to switch between the two. Thank you!

File: blockchain/chain_sync/src/sync.rs
Comment (austinabell) [reply blockchain/chain_sync/src/sync.rs:71]: I just mean wrapping the variables other than the blockstore in the mutex, so you don't wrap the whole struct (and would restrict blockstore access even if just one of the other variables are locked)

File: node/rpc/src/lib.rs
Comment (AshantiMutinta) [reply node/rpc/src/lib.rs:210]: That is true, maybe a warn might be better here?

Comment (ec2) [reply node/rpc/src/lib.rs:210]: Sounds good to me

**Comment by AshantiMutinta** [comment]
> Gunna go much more in depth tomorrow. My cloud lotus node is currently syncing.
> Could you also show what commands you run to do some of the testing? What I do to lotus nodes is
> 
>     1. Run `wscat -c 127.0.0.1:1234/rpc/v0`
> 
>     2. Send requests like:` {"jsonrpc":"2.0","method":"Filecoin.ChainNotify","params":[],"id":1}`

I would use the same commands :)

**Comment by austinabell** [comment]
You committed the grpc protobuf submodule by the way (and I assume unintended)

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:214]: Won't this just immediately time out? what is the purpose of this?

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:201]: This just errors when all receivers have been dropped, are you sure you want to error here?

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:667]: Is this untagged or tagged with PascalCase? These are contradictory. How are you checking the format?

**Review by austinabell — COMMENTED** [review]
I still do believe that only the variables that need to be mutated across threads in the chain store should be wrapped in a mutex, you are adding locks around a lot of logic that doesn't need it and performance and/or deadlocks might be a problem

**Comment by AshantiMutinta** [comment]
Mentioned on the slack chat that this will probably be done on Sunday. Put this up like this in case you needed a temporary merge

File: blockchain/chain/src/store/chain_store.rs
Comment (AshantiMutinta) [reply blockchain/chain/src/store/chain_store.rs:214]: This is for cases where the pending future doesn't change which is flostreams design. This will timeout when future in the foreach hasn't  been resolved

File: blockchain/chain/src/store/chain_store.rs
Comment (AshantiMutinta) [reply blockchain/chain/src/store/chain_store.rs:201]: No harm in leaving it aside from if there is any other problem into sending to a channel

File: blockchain/chain/src/store/chain_store.rs
Comment (AshantiMutinta) [reply blockchain/chain/src/store/chain_store.rs:667]: This was a mistake, it is supposed to be tagged

Comment (austinabell) [reply blockchain/chain/src/store/chain_store.rs:667]: But the enum variants are not pascal cased in Lotus? They are all lower case from what I can see

File: node/forest_libp2p/src/service.rs
Comment (nannick) [inline node/forest_libp2p/src/service.rs:262]: Delete comment

Comment (austinabell) [reply node/forest_libp2p/src/service.rs:262]: This just shouldn't be commented out (imo, if anyone disagrees please say why)

Comment (ec2) [reply node/forest_libp2p/src/service.rs:262]: I'd say uncomment it.

**Comment by AshantiMutinta** [comment]
This PR uses a forked version of JSONRPC_V2

https://github.com/AshantiMutinta/jsonrpc-v2

**Comment by austinabell** [comment]
> This PR uses a forked version of JSONRPC_V2
> 
> https://github.com/AshantiMutinta/jsonrpc-v2

If this is planning to stay as such for a non-trivial amount of time, can you fork it to the ChainSafe org and give a few of us privileges in case we need to change something?

Also, did you end up trying to use jsonrpsee for this?

**Comment by AshantiMutinta** [comment]
> 
> 
> > This PR uses a forked version of JSONRPC_V2
> > https://github.com/AshantiMutinta/jsonrpc-v2
> 
> If this is planning to stay as such for a non-trivial amount of time, can you fork it to the ChainSafe org and give a few of us privileges in case we need to change something?
> 
> Also, did you end up trying to use jsonrpsee for this?

I thought the jsonrpsee was still in it's fairly early stages. It shouldn't be too hard to switch if we really want it though, I think for the time being since we want the storage miner in, forking is a good alternative. 

I would switch in a subsequent PR if we really prefer it though

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:748]:
```suggestion
    #[derive(Serialize)]
```

File: blockchain/blocks/src/tipset.rs
Comment (austinabell) [inline blockchain/blocks/src/tipset.rs:274]:
```suggestion
    #[derive(Deserialize, Serialize)]
```

File: blockchain/blocks/src/tipset.rs
Comment (austinabell) [inline blockchain/blocks/src/tipset.rs:279]:
```suggestion
    #[derive(Serialize)]
```

File: forest/src/daemon.rs
Comment (austinabell) [inline forest/src/daemon.rs:131]: This subscriber is cloned here, but never polled off of, so this will stall eventually (need to be certain no subscriber _doesn't_ get polled

File: node/rpc/src/lib.rs
Comment (austinabell) [inline node/rpc/src/lib.rs:56]: If anything, this should be an `Arc<Publisher<..>>` where if you need to get something to subscribe to it, you use that to create a new subscriber

File: node/rpc/src/lib.rs
Comment (austinabell) [inline node/rpc/src/lib.rs:264]: be careful cloning the subscribers, this would probably change if you change the previous comments, but the subscriber that's being cloned isn't being polled off

**Review by austinabell — CHANGES_REQUESTED** [review]

File: node/rpc/src/lib.rs
Comment (AshantiMutinta) [reply node/rpc/src/lib.rs:319]: Hmm I could switch that around, annoying because it only accepts static strings but easy to change since we have forks

File: node/rpc/src/lib.rs
Comment (AshantiMutinta) [reply node/rpc/src/lib.rs:386]: I think this is one of those cases where we have unrecoverable errors. I looked at it like we are the ones doing the serialization and at this point the serialization should always pass from our side since we are in charge of the implementation

File: blockchain/chain/src/store/chain_store.rs
Comment (AshantiMutinta) [reply blockchain/chain/src/store/chain_store.rs:770]: The publisher is needed here because you need to make sure you are iterating from the first the beginning of the channel in situations where RPCState is being shared amongst processes

File: blockchain/chain/src/store/chain_store.rs
Comment (AshantiMutinta) [reply blockchain/chain/src/store/chain_store.rs:247]: This came in before your changes for the ChainSyncer, they were lock versions of the functions. They should be removed

File: node/rpc/src/lib.rs
Comment (AshantiMutinta) [reply node/rpc/src/lib.rs:56]: The reason why this is a subscribe is that it is a one time usage just for the purpose of the sub_head_changes function. We are only ever using this to poll from the subscribe channel to iterate our HeadChangeJson.

Having a publisher return subscribers will mean that publisher will have to be behind a Arc<Lock which also carries overhead

File: forest/src/daemon.rs
Comment (AshantiMutinta) [reply forest/src/daemon.rs:131]: What do you mean? It's being polled in the sub_head_changes. Maybe im missing something?

File: blockchain/chain/src/store/chain_store.rs
Comment (AshantiMutinta) [reply blockchain/chain/src/store/chain_store.rs:328]: remnants of debugging I think, will remove this

File: node/rpc/src/lib.rs
Comment (AshantiMutinta) [reply node/rpc/src/lib.rs:43]: Changed to &'a str instead

File: node/rpc/src/lib.rs
Comment (austinabell) [reply node/rpc/src/lib.rs:315]: why don't you use the `Full` error variant, and then no need to change?

File: node/rpc/src/lib.rs
Comment (austinabell) [reply node/rpc/src/lib.rs:381]: yeah, that might be correct, but I don't see the negative in handling to avoid crashing or to future proof if something else is changed

File: node/rpc/src/lib.rs
Comment (austinabell) [reply node/rpc/src/lib.rs:43]: it doesn't need a lifetime though?

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [reply blockchain/chain/src/store/chain_store.rs:770]: Maybe I'm confused about how this is being published to multiple threads, I can look into this later.

File: forest/src/daemon.rs
Comment (austinabell) [reply forest/src/daemon.rs:131]: but that's a cloned subscriber?

File: node/rpc/src/lib.rs
Comment (austinabell) [reply node/rpc/src/lib.rs:56]: but this RpcState is holding this subscriber that never gets polled, as far as I can see (can you point to where it's being handled?).

The only reason you would hold the publisher is so that you'd be able to subscribe without holding a dead subscriber. I do see the overhead concern with this, especially since the subscribe would require a write lock for some reason, but as of now that would be better than the node stalling after this subscriber buffer fills

File: node/rpc/src/lib.rs
Comment (AshantiMutinta) [reply node/rpc/src/lib.rs]: static will mean those string will exist the whole time the program is running. &'a in this case will only limit to the lifetime of the enum which is better imo

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:712]: These make no sense to me, why would you ever need this that you wouldn't be able to do with a match. (I see where it's used, but it doesn't seem clear to me why

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:757]: why is a cancel event published on a pubsub type channel? When would you ever want to send something to cancel all subscriptions? I'm really confused on this pattern, or how it would be used, if you don't mind elaborating the intention

File: blockchain/chain_sync/src/sync.rs
Comment (austinabell) [inline blockchain/chain_sync/src/sync.rs:349]: there is nothing async in this function?

File: forest/src/daemon.rs
Comment (austinabell) [inline forest/src/daemon.rs:73]:
```suggestion
    let subscriber = publisher.write().await.subscribe();
```
This will deref automatically

File: node/rpc/src/lib.rs
Comment (austinabell) [inline node/rpc/src/lib.rs:17]:
```suggestion
use async_std::sync::{RwLock, Sender, Arc};
use async_std::task::{self,JoinHandle};
```
Can you group imports like this?

File: node/rpc/src/lib.rs
Comment (austinabell) [inline node/rpc/src/lib.rs:25]:
```suggestion
use flo_stream::{MessagePublisher, Publisher, Subscriber};
```

File: node/rpc/src/lib.rs
Comment (austinabell) [inline node/rpc/src/lib.rs:264]:
```suggestion
        let subscriber = events_pubsub.write().await.subscribe();
```

File: node/rpc/src/lib.rs
Comment (austinabell) [inline node/rpc/src/lib.rs:412]: just a thought, can this not just be `&mut WsSink` and just await a write lock on the function call so you don't need to clone the arc when you are awaiting the response anyway?

File: node/rpc/src/lib.rs
Comment (austinabell) [reply node/rpc/src/lib.rs:450]: ^^ What was the reason why this is resolved? Intentionally doing for what reason?

File: node/rpc/src/lib.rs
Comment (austinabell) [inline node/rpc/src/lib.rs:454]: why does this need to be cloned?

File: node/rpc/src/state_api.rs
Comment (austinabell) [inline node/rpc/src/state_api.rs:289]: I can't comment on previous, but won't this be incorrect serialization because it's going to serialize as bytes? Default json serialization should be string?

File: blockchain/chain/src/store/chain_store.rs
Comment (AshantiMutinta) [reply blockchain/chain/src/store/chain_store.rs:757]: The cancel even t is not sent to a publish. The channel that is done with the sub_head_changes is sent to this function with the appropriate id and the sub_head_changes method that matches up with this ID is cancelled. They can be a of sub_head_changes running at the same time. So this doesn't publish but checks if an event with the subscription ID has been passed through

File: blockchain/chain_sync/src/sync.rs
Comment (AshantiMutinta) [reply blockchain/chain_sync/src/sync.rs:349]: Probably a remnant from an old refactor

File: blockchain/chain/src/store/chain_store.rs
Comment (AshantiMutinta) [reply blockchain/chain/src/store/chain_store.rs:712]: This is for the purpose of using it in filter closures. I want to be able to say task_cancel().map(|s|//operation goes here), I find it match cleaner than adding a match in a closure

File: blockchain/chain/src/store/chain_store.rs
Comment (AshantiMutinta) [reply blockchain/chain/src/store/chain_store.rs:65]: I thought it was suggested the first time that the RPCState should just obtain mutable values that it needs to refrain from potentially blocking the blockstore. 

These both need to be in Sync Channel and RPCState, this is straying away from the block_store since we do not want to block that. It is a "taking what we need approach"

File: node/rpc/src/lib.rs
Comment (AshantiMutinta) [reply node/rpc/src/lib.rs:410]: The downside of that is the lock will be obtained for the whole lifetime of the function rather than just the single line that needs the send but I can change

Comment (austinabell) [reply node/rpc/src/lib.rs:409]: Ah ya, I guess I meant a ref to the RwLock, to avoid the arc clone

File: node/rpc/src/lib.rs
Comment (AshantiMutinta) [reply node/rpc/src/lib.rs:451]: That was a mistake, will remove that

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:174]: You really shouldn't be describing these functions with the field type. Reason being is that if that type ever changes, you need to either have to rename the function and make breaking changes, or leave and have the name be misleading.

I suggest taking off the `_rwlock` suffixes, and switching the `_arc` suffix to `_cloned`

File: blockchain/chain/Cargo.toml
Comment (austinabell) [inline blockchain/chain/Cargo.toml:29]:
```suggestion
futures = "0.3.5"
```

**Review by austinabell — APPROVED** [review]
LGTM pending @ec2's checks

**Review by ec2 — APPROVED** [review]
forgot the hit approve lgtm

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:744]: Also, why is this type in the `headchange_json` mod?

File: blockchain/chain/src/store/chain_store.rs
Comment (austinabell) [inline blockchain/chain/src/store/chain_store.rs:732]: Also, did you mean to annotate the headchange type here to use the `HeadChangeJson` serialization?

**Pull request merged**
